packages/dev-server-hmr/src/HmrPlugin.ts

```diff
@@ -0,0 +1,293 @@
+import type {
+  Plugin,
+  WebSocketsManager,
+  Logger,
+  WebSocketData,
+  ServerStartParams,
+} from '@web/dev-server-core';
+import WebSocket from 'ws';
+import type { Context } from 'koa';
+import { hmrClientScript } from './hmrClientScript';
+import { posix as pathUtil } from 'path';
+
+export interface HmrReloadMessage {
+  type: 'hmr:reload';
+}
+
+export interface HmrUpdateMessage {
+  type: 'hmr:update';
+  url: string;
+}
+
+export interface HmrAcceptMessage extends WebSocketData {
+  type: 'hmr:accept';
+  id: string;
+}
+
+export type HmrMessage = HmrReloadMessage | HmrUpdateMessage;
+
+export type HmrClientMessage = HmrAcceptMessage;
+
+export interface HmrModule {
+  dependencies: Set<string>;
+  dependents: Set<string>;
+  hmrAccepted: boolean;
+  hmrEnabled: boolean;
+}
+
+export const NAME_HMR_CLIENT_IMPORT = '/__web-dev-server__/hmr.js';
+
+/**
+ * Dev server plugin to provide hot module reloading
+ */
+export class HmrPlugin implements Plugin {
+  name = 'hmr';
+  injectWebSocket = true;
+
+  protected _dependencyTree: Map<string, HmrModule> = new Map();
+  protected _webSockets?: WebSocketsManager;
+  protected _logger?: Logger;
+
+  /** @inheritDoc */
+  async serverStart({ webSockets, fileWatcher, logger }: ServerStartParams) {
+    if (!fileWatcher) {
+      throw new Error('Cannot use HMR when watch mode is disabled.');
+    }
+
+    if (!webSockets) {
+      throw new Error('Cannot use HMR when web sockets are disabled.');
+    }
+
+    this._webSockets = webSockets;
+    this._logger = logger;
+
+    webSockets.on('message', ({ webSocket, data }) => this._onMessage(webSocket, data));
+    fileWatcher.on('change', path => this._onFileChanged(path));
+    fileWatcher.on('unlink', path => this._onFileChanged(path));
+
+    this._logger?.debug('[hmr] Listening for HMR messages');
+  }
+
+  /** @inheritDoc */
+  async serve(context: Context) {
+    // Someone is requesting the injected client script
+    if (context.path === NAME_HMR_CLIENT_IMPORT) {
+      if (!this._webSockets) {
+        return;
+      }
+      return hmrClientScript(this._webSockets);
+    }
+
+    // We are serving a new file or it has changed, so clear all the
+    // dependencies we previously tracked (if any).
+    this._clearDependencies(context.path);
+    this._logger?.debug(`[hmr] Cleared dependency tree cache of ${context.path}`);
+  }
+
+  /** @inheritDoc */
+  async transformImport({ source, context }: { source: string; context: Context }) {
+    // Can't possibly handle computed dynamic imports
+    if (source.includes('${')) {
```

Comment:
this feels like hackery but we need to draw the line somewhere on computed imports. so i just went with "if any dynamic import is computed, do not handle it".

example:

```ts
import(`${foo}/${bar}`).then(...); // wouldn't be handled by HMR, no clue what you imported
```

Reply:
We can add an indicator to the `transformImport` function to indicate it's a dynamic import.